desktop/cmp/grouping/GroupingChooser.ts

```diff
@@ -107,35 +230,53 @@ export const [GroupingChooser, groupingChooser] = hoistCmp.withFactory<GroupingC
                         e?.target &&
                         !targetIsControlButtonOrPortal(e.target)
                     ) {
-                        model.commitPendingValueAndClose();
+                        impl.commitPendingValueAndClose();
                     }
                 }
             })
         });
     }
 });
 
+interface WithLocalModelProps extends HoistProps<GroupingChooserModel> {
+    impl: GroupingChooserLocalModel;
+}
+
 //------------------
 // Editor
 //------------------
-const editor = hoistCmp.factory<GroupingChooserModel>({
-    render({popoverWidth, popoverMinHeight, popoverTitle, emptyText}) {
+
+interface EditorProps
+    extends WithLocalModelProps,
+        Pick<
+            GroupingChooserProps,
+            'emptyText' | 'popoverMinHeight' | 'popoverWidth' | 'popoverTitle'
+        > {
+    impl: GroupingChooserLocalModel;
+}
+
+const editor = hoistCmp.factory<EditorProps>({
+    render({popoverWidth, popoverMinHeight, popoverTitle, emptyText, impl}) {
         return panel({
             width: popoverWidth,
             minHeight: popoverMinHeight,
             items: [
                 div({className: 'xh-popup__title', item: popoverTitle, omit: !popoverTitle}),
-                dimensionList({emptyText}),
-                addDimensionControl(),
+                dimensionList({emptyText, impl}),
+                addDimensionControl({impl}),
                 filler()
             ]
         });
     }
 });
 
-const dimensionList = hoistCmp.factory<GroupingChooserModel>({
-    render({model, emptyText}) {
-        if (isEmpty(model.pendingValue)) {
+interface DimensionListProps extends WithLocalModelProps {
+    emptyText: string;
+}
+
+const dimensionList = hoistCmp.factory<DimensionListProps>({
+    render({model, emptyText, impl}) {
```

Comment:
same as above

Reply:
Suggesting to move this into a shared impl directory? Can't really do that unfortunately as it's child components use platform-specific components

Reply:
discussed with John -- there was a misunderstanding -- just mean that we should be passing the state down to the local components as `dimensionList({imodel: impl})`.   That way they have typed access to everything.